Add 'Nullable' option to the build property page

drewnoakes opened the pull request:
Fixes #4058.

This option is only available when the language version is 8.0 or greater.

---

### netcoreapp3.1 (LangVersion 8)

![image](https://user-images.githubusercontent.com/350947/72721113-71b24900-3bcf-11ea-9ac3-2737380ba1e9.png)

The dropdown has the following options:

![image](https://user-images.githubusercontent.com/350947/72721168-87c00980-3bcf-11ea-92e4-8e8d7488fe56.png)

---

### net48 (LangVersion 7.3)

The nullable control is disabled:

![image](https://user-images.githubusercontent.com/350947/72770664-a9aea000-3c52-11ea-95c7-dec54933481f.png)

davkean commented:
I think we should consider coming up with a way for this property to visible but disabled when changing the TFM will cause it to be enabled.

- What happens if I choose the Application tab, change from NET Core 3.0 and then choose this tab?

davkean commented:
This is the help for that page: https://docs.microsoft.com/en-us/visualstudio/ide/reference/build-page-project-designer-csharp?f1url=https%3A%2F%2Fmsdn.microsoft.com%2Fquery%2Fdev16.query%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(cs.ProjectPropertiesBuild)%26rd%3Dtrue&view=vs-2019. We should document it there.

drewnoakes commented:
Changing it to disable rather than hide is straightforward to do, but I felt it would make the UI more confusing for users.

Currently the state of this UI is not updated based on changes to TFM (via either UI or file on disk). The page needs to be reloaded. I looked at `PropertyControlData` for hints on how to respond to such changes but didn't see anything.

We could justify sticking help links on every UI element on the page. Does this one deserve one just because it's new?

Once the UI here is finalised I'll update the docs.

davkean commented:
We don't hide elements in other cases where changing the TFM would enable it - for example, Prefer 32-bit. Can you see how it reacts when you change from 4.0 -> 4.5 in legacy?

davidwengier commented:
> Does this one deserve one just because it's new?

I think this one deserves something because "Enable", "Warnings" and "Annotations" don't mean much on their own.

drewnoakes commented:
Updated to respond to changes to `LangVersion`/`TargetFramework`, either through the UI or project file.

Also changed to disable rather than hide the UI.

drewnoakes commented:
> I think this one deserves something because "Enable", "Warnings" and "Annotations" don't mean much on their own.

I think you could argue this about any of the options you weren't familiar with. For example, what does the `TRACE` constant mean? What's a constant? What does warning level '4' mean?

The property pages support <kbd>F1</kbd> help, though it's not very discoverable and is currently broken (I filed #5815).

drewnoakes commented:
Documentation updated in https://github.com/MicrosoftDocs/visualstudio-docs-pr/pull/6024.

tmeschter commented on src/Microsoft.VisualStudio.Editors/Resources/Microsoft.VisualStudio.Editors.Designer.resx:
```diff
@@ -1970,6 +1970,18 @@ Do you want to update the value in the .settings file?</value>
+  </data>
+  <data name="PPG_BuildSettings_Nullable_Annotations" xml:space="preserve">
+    <value>Annotations</value>
+  </data>
```
By placing these values in .resx file we're choosing to localize them. Is that what you want, or would it make more sense to simply use the literal values understood by the compiler?

In the case of the latter you would hard-code them in the source and use the lower-case forms.

davkean replied:
The drop down should be localized, what we persist in the project file shouldn't be.

drewnoakes replied:
I considered using the compiler's strings and ended up following the example set by the advanced build properties page (where I initially added this new UI before moving it to the regular build properties page).

![image](https://user-images.githubusercontent.com/350947/72856758-a0cfd400-3d0f-11ea-9392-65fa919936fb.png)

![image](https://user-images.githubusercontent.com/350947/72856742-98779900-3d0f-11ea-8e33-4c6aaa3b021c.png)

davkean replied:
Yes, and these were localized in response to a bug.

davkean commented:
The "Advanced Build Options" page is also configuration-dependent. I think you can argue this for lots of options on that page, unsafe code, Prefer 32-bit, etc.

drewnoakes commented:
@tmeschter 

> If you want it to be configuration-independent, the best option is the Advanced Build Options dialog.

To be clear, in this PR the setting _behaves_ in a configuration-independent manner due to `HasConfigurationCondition="False"` on its data source.

The question is whether this behaviour is surprising to the user or not.

@davkean @davidwengier and I discussed placement on Monday and decided the setting was not really 'advanced', and that it deserved a more prominent position in the UI. So I moved it from the advanced build prop page to the regular build prop page.

But looking at it now I see that every other property on that page is actually configuration-dependent.

I agree that nullable shouldn't be configuration-dependent. It's a shame to hide the option further down, but on balance I think you're right so will move it back 🤦‍♂

drewnoakes commented:
_Actually_ it turns out that every setting on the _advanced_ property page is configuration dependent too. So now I'm less sure that moving it does anything useful.

When language version was editable, it would have been the exception on all these pages.

tmeschter commented:
Sorry, I missed `HasConfigurationCondition="False"` on the data source--I'm so used to thinking about these pages through the lens of the old project system. :-) I agree that with that issue out of the way, the Build page is a much better choice than the Advanced dialog. Approved.
